liquidhaskell-boot/src/Language/Haskell/Liquid/Bare/Axiom.hs

```diff
@@ -154,23 +156,63 @@ strengthenSpecWithMeasure sig env actualV qPretended =
-getReflectDefs src sig spec = findVarDefType cbs sigs <$> xs
+getReflectDefs src sig spec env modName = go initialXs initialDefined []
   where
     sigs                    = gsTySigs sig
-    xs                      = S.toList (Ms.reflects spec)
+    initialXs               = S.toList (Ms.reflects spec)
     cbs                     = _giCbs src
-
-findVarDefType :: [Ghc.CoreBind] -> [(Ghc.Var, LocSpecType)] -> LocSymbol
-               -> (LocSymbol, Maybe SpecType, Ghc.Var, Ghc.CoreExpr)
-findVarDefType cbs sigs x = case findVarDefMethod (val x) cbs of
+    initialDefined          = M.empty
+
+    go :: [LocSymbol] -> M.HashMap LocSymbol Ghc.Var 
+       -> [(LocSymbol, Maybe SpecType, Ghc.Var, Ghc.CoreExpr)] 
+       -> [(LocSymbol, Maybe SpecType, Ghc.Var, Ghc.CoreExpr)]
+    go [] _ acc = acc
+    go xs defined acc = if null found
+         then case newXs of
+                [] -> acc
+                x:_ -> Ex.throw . mkError x $
+                  "Not found in scope nor in the amongst these variables: " ++
+                    foldr (\x acc -> acc ++ " , " ++ show x) "" newDefined
+         else go newXs newDefined newAcc
+      where
+        results   = findVarDefType cbs sigs env modName defined <$> xs
+        found     = Mb.catMaybes results
+        newAcc    = acc ++ found
+        newDefined = foldl addFreeVarsToMap defined found
+        newXs     = [x | (x, Nothing) <- zip xs results]
+
+    addFreeVarsToMap defined (_, _, _, expr) =
+      let freeVarsSet = getAllFreeVars expr
+          newBindings = M.fromList [(Bare.varLocSym var, var) | var <- S.toList freeVarsSet]
+      in M.union defined newBindings
+
+    getAllFreeVars = S.fromList . Ghc.exprSomeFreeVarsList (const True)
+
+findVarDefType :: [Ghc.CoreBind] -> [(Ghc.Var, LocSpecType)] -> Bare.Env -> ModName
+               -> M.HashMap LocSymbol Ghc.Var -> LocSymbol
+               -> Maybe (LocSymbol, Maybe SpecType, Ghc.Var, Ghc.CoreExpr)
+findVarDefType cbs sigs env modName defs x = case findVarDefMethod (val x) cbs of
   -- YL: probably ok even without checking typeclass flag since user cannot
   -- manually reflect internal names
   Just (v, e) -> if GM.isExternalId v || isMethod (F.symbol x) || isDictionary (F.symbol x)
-                   then (x, val <$> lookup v sigs, v, e)
+                   then Just (x, val <$> lookup v sigs, v, e)
                    else Ex.throw $ mkError x ("Lifted functions must be exported; please export " ++ show v)
-  Nothing     -> Ex.throw $ mkError x $ show (val x) ++ " is not in scope"
+  Nothing     -> do
+    var <- Bare.maybeResolveSym env modName "findVarDefType" qSym <|> M.lookup qSym defs
+    let info = Ghc.idInfo var
+    let unfolding = getExpr . Ghc.realUnfoldingInfo $ info
+    case unfolding of
+      Just e -> Just (x, val <$> lookup var sigs, var, e)
+      _ -> Ex.throw $ mkError x "Symbol exists but is not defined in the current file, and no unfolding is available in the interface files"
```

Comment:
Why is this function returning a Maybe if all the branches give either Just or throw an error?

Reply:
Well, actually there's one possible place for returning Nothing:
```
var <- Bare.maybeResolveSym env modName "findVarDefType" qSym <|> M.lookup qSym defs
```
And that's key because of the way it's used in `getReflectDefs`. We don't want to throw an error (yet) if the variable is not defined. Because it might be defined afterward in the fixpoint computation

Reply:
(but I should probably add a comment to say it)

Reply:
Done in a8a111a42c324387b8cea0c9cd20b2295d0ec83d